### PR TITLE
<random>: Fix mersenne_twister_engine special case behavior

### PR DESCRIPTION
**Pull request by statementreply**
- Fixes `mersenne_twister_engine` random number generation when _m_ = _n_ or _n_ = 1. [[rand.eng.mers]/1](https://wg21.link/rand.eng.mers#1):
  > [...] all subscripts applied to _X_ are to be taken modulo _n_.
- Fixes `mersenne_twister_engine` seeding when input is degenerate.
- Also removes the trailing space of text output.

### CODE REVIEW

File: stl/inc/random [hunk 1]
```diff
@@ -1116,27 +1118,43 @@ protected:
     _Post_satisfies_(this->_Idx == 0)
 
         void _Refill_lower() { // compute values for the lower half of the history array
-        size_t _Ix;
-        for (_Ix = 0; _Ix < _Nx - _Mx; ++_Ix) { // fill in lower region
-            _Ty _Tmp       = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[_Ix + _Nx + 1] & _LMSK);
-            this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix + _Nx + _Mx];
-        }
+        if constexpr (_M_mod_n > 0) {
```
Comment (miscco) [inline stl/inc/random:1121]: I am wondering, whether it would be clearer to flip and check for `_M_mod_n == 0`

File: stl/inc/random [hunk 2]
```diff
@@ -1116,27 +1118,43 @@ protected:
...
+            size_t _Ix;
+            for (_Ix = 0; _Ix < _Nx - _One_mod_n; ++_Ix) { // fill in values
+                _Ty _Tmp       = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[_Ix + _Nx + _One_mod_n] & _LMSK);
+                this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix + _Nx];
+            }
+
+            if constexpr (_One_mod_n > 0) { // fill in _Ax[_Nx - 1]
+                _Ty _Tmp       = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[0] & _LMSK);
+                this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix + _Nx];
+            }
```
Comment (miscco) [inline stl/inc/random:1147]: I am missing any relation to `_One_mod_n`

The other clause has `this->_Ax[_M_mod_n - 1]` at the end. Is this one here correct?

**Review by miscco — COMMENTED** [review]
I am not really that deep into mersenne twister so some superficial comments that made me scratch my head

File: stl/inc/random (hunk 2 above)
Comment (statementreply) [reply stl/inc/random:1147]: `_Ix` is `_Nx - 1` here, thus
- `_Ix + _Nx` = `_Ix + _Nx + _M_mod_n` = `(_Ix + _Nx + _M_mod_n) % (_Nx * 2)` in this clause,
- `_M_mod_n - 1` = `_Ix - _Nx + _M_mod_n` = `(_Ix + _Nx + _M_mod_n) % (_Nx * 2)` in the other clause,

so both are correct.

I should probably substitute `_Ix` for `_Nx - 1` in both clauses.

File: stl/inc/random [hunk 3]
```diff
@@ -1116,27 +1118,45 @@ protected:
...
-            this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix + _Nx + _Mx];
-        }
+        if constexpr (_M_mod_n == 0) {
+            for (size_t _Ix = 0; _Ix < _Nx - _One_mod_n; ++_Ix) { // fill in values
+                const _Ty _Tmp = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[_Ix + _Nx + _One_mod_n] & _LMSK);
+                this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix + _Nx + _M_mod_n];
+            }
+
+            if constexpr (_One_mod_n == 1) { // fill in _Ax[_Nx - 1]
+                constexpr size_t _Ix = _Nx - _One_mod_n;
```
Comment (miscco) [inline stl/inc/random:1128]: This is the end condition of the above loop, so you could pull that up and name is something differently like `constexpr size_t _Last_ix = _Nx - _One_mod_n;`

File: stl/inc/random [hunk 4]
```diff
@@ -1116,27 +1118,45 @@ protected:
...
+                // fill in upper region (avoids modulus operation)
+                const _Ty _Tmp = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[_Ix + _Nx + _One_mod_n] & _LMSK);
+                this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix - _Nx + _M_mod_n];
+            }
 
-        for (; _Ix < _Nx - 1; ++_Ix) { // fill in upper region (avoids modulus operation)
-            _Ty _Tmp       = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[_Ix + _Nx + 1] & _LMSK);
-            this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix - _Nx + _Mx];
+            if constexpr (_One_mod_n == 1) { // fill in _Ax[_Nx - 1]
+                constexpr size_t _Ix = _Nx - _One_mod_n;
```
Comment (miscco) [inline stl/inc/random:1146]: Same here

File: stl/inc/random [hunk 5]
```diff
@@ -1116,27 +1118,48 @@ protected:
...
+                this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix + _Nx + _M_mod_n];
+            }
+
+            for (size_t _Ix = _Wrap_bound_m; _Ix < _Wrap_bound_one; ++_Ix) {
+                // fill in upper region (avoids modulus operation)
+                const _Ty _Tmp = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[_Ix + _Nx + _One_mod_n] & _LMSK);
+                this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix - _Nx + _M_mod_n];
+            }
+
+            if constexpr (_One_mod_n == 1) { // fill in _Ax[_Nx - 1]
```
Comment (miscco) [inline stl/inc/random:1148]: This says `// fill in _Ax[_Nx - 1]` but below you write `this->_Ax[_Ix]`, which is a bit confusing, same above

Comment (StephanTLavavej) [reply stl/inc/random:1148]: Good catch, but I think this is OK - the commented conditions are when `_One_mod_n == 1`, so `_Ax[_Ix]` is always `_Ax[_Nx - 1]`. If I understand @statementreply's intent correctly, `_Ax[_Ix]` is being used to highlight the similar structure of these assignments. (For similar reasons, `_One_mod_n` is repeated even when it's known to be `1`.)

**Review by StephanTLavavej — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for fixing these 3 bugs - clearly chosen because 3 is both a Mersenne prime exponent and a Mersenne prime itself! :joy_cat:

**Pull request merged**